File: NOIP/Repecharge/NOIP2010/提高组参考程序/C++/prison.cpp
#include <iostream>
#include <fstream>

using namespace std;
struct edge{
	int v,w;
	edge *next;
};const int maxn=20000,maxm=100000;
int n,m,ans;
edge *g[maxn+1];
int a[maxm+1];
int q[maxm*10];
int color[maxn+1];

void add(int u,int v,int w)
{
	edge* p;
	p=new edge;
	p->v=v;p->w=w;p->next=g[u];
	g[u]=p;
}

void qsort(int l,int r)
{
	int i,j,k,tmp;
	i=l;j=r;k=a[(i+j)/2];
	while(i<=j){
		while(a[i]<k) i++;
		while(a[j]>k) j--;
		if(i<=j){
			tmp=a[i];a[i]=a[j];a[j]=tmp;
			i++;j--;
		}
	}
	if(i<r) qsort(i,r);
	if(j>l) qsort(l,j);
}

void init()
{
	ifstream infile("prison.in",ios::in);
	memset(g,0,sizeof(g));
	int u,v,w;
	infile>>n>>m;
	for(int i=1;i<=m;i++){
		infile>>u>>v>>w;
		add(u,v,w);add(v,u,w);
		a[i]=w;
	}
	qsort(1,m);
	infile.close();
}

void over()
{
	ofstream outfile("prison.out",ios::out);
	outfile<<ans<<endl;
	outfile.close();
}

bool bfs(int weight)
{
	int f,r,u,v,w;
	edge* p;
	memset(color,0,sizeof(color));
	for(int k=1;k<=n;k++) if(color[k]==0){
		f=1;r=0;
		r++;q[r]=k;color[k]=1;
		while(f<=r){
			u=q[f];f++;
			p=g[u];
			while(p!=NULL){
				v=p->v;w=p->w;
				if(w>weight){
					if((color[v]==color[u])&&(color[v]!=0))
						return false;
					else if(color[v]==0){
						r++;q[r]=v;
						color[v]=(color[u]%2)+1;
					}
				}
				p=p->next;
			}
		}
	}
	return true;
}
int main()
{
	int l,r,mid;
	init();
	l=1;r=m;
	while(l<=r){
		mid=(l+r)/2;
		if(bfs(a[mid])) r=mid-1;
		else l=mid+1;
	}
	if(r<=0) ans=0;
	else ans=a[l];
	over();
	return 0;
}

File: NOIP/Final/NOIP2010/SeniorRefCode/Pascal/prison.pas
program prison(input,output);
  const
    maxn=20000;
    maxm=100000;
  type
    link=^edge;
    edge=record
      v,w:longint;
      next:link;
    end;
  var
    n,m,ans:longint;
    g:array[1..maxn] of ^edge;
    a:array[1..maxm] of longint;
    q:array[1..maxm*10] of longint;
    color:array[1..maxn] of longint;

  procedure add(u,v,w:longint);
    var
      p:link;
    begin
      new(p);
      p^.v:=v;p^.w:=w;p^.next:=g[u];
      g[u]:=p;
    end;

  procedure init;
    var
      i,u,v,w:longint;
    begin
      fillchar(g,sizeof(g),0);
      assign(input,'prison.in');reset(input);
      readln(n,m);
      for i:=1 to m do begin
        readln(u,v,w);
        add(u,v,w);
        add(v,u,w);
        a[i]:=w;
      end;
      close(input);
    end;

  procedure over;
    begin
      assign(output,'prison.out');rewrite(output);
      writeln(ans);
      close(output);
    end;

  procedure qsort(l,r:longint);
    var
      i,j,k,tmp:longint;
    begin
      i:=l;j:=r;k:=a[(i+j) shr 1];
      while i<=j do begin
        while a[i]<k do inc(i);
        while a[j]>k do dec(j);
        if i<=j then begin
          tmp:=a[i];a[i]:=a[j];a[j]:=tmp;
          inc(i);dec(j);
        end;
      end;
      if i<r then qsort(i,r);
      if l<j then qsort(l,j);
    end;

  function bfs(weight:longint):boolean;
    var
      f,r,k,u,v,w:longint;
      p:link;
    begin
      fillchar(color,sizeof(color),0);
      for k:=1 to n do if color[k]=0 then begin
        f:=1;r:=0;
        inc(r);q[r]:=k;color[k]:=1;
        while f<=r do begin
          u:=q[f];inc(f);
          p:=g[u];
          while p<>nil do begin
            v:=p^.v;w:=p^.w;
            if w>weight then begin
              if (color[v]<>0)and(color[v]=color[u]) then
                exit(false)
              else if (color[v]=0) then begin
                color[v]:=color[u] mod 2+1;
                inc(r);q[r]:=v;
              end;
            end;
            p:=p^.next;
          end;
        end;
      end;
      exit(true);
    end;

  procedure main;
    var
      l,r,mid:longint;
    begin
      qsort(1,m);
      l:=1;r:=m;
      while l<=r do begin
        mid:=(l+r) shr 1;
        if bfs(a[mid]) then r:=mid-1
        else l:=mid+1;
      end;
      if r<=0 then ans:=0
      else ans:=a[l];
    end;

  begin
    init;
    main;
    over;
  end.
